// vuepress-theme-hope/lib/client/modules/info/components/PageInfo.js
import { defineComponent, resolveComponent, h } from "vue";
import { usePure } from "@theme-hope/composables/index.js";
import AuthorInfo from "@theme-hope/modules/info/components/AuthorInfo.js";
import CategoryInfo from "@theme-hope/modules/info/components/CategoryInfo.js";
import DateInfo from "@theme-hope/modules/info/components/DateInfo.js";
import PageViewInfo from "@theme-hope/modules/info/components/PageViewInfo.js";
import ReadingTimeInfo from "@theme-hope/modules/info/components/ReadingTimeInfo.js";
import TagInfo from "@theme-hope/modules/info/components/TagInfo.js";
import OriginalInfo from "@theme-hope/modules/info/components/OriginalMark.js";
import WordInfo from "@theme-hope/modules/info/components/WordInfo.js";
import "balloon-css/balloon.css";
import "../styles/page-info.scss";
export default defineComponent({
    name: "PageInfo",
    components: {
        AuthorInfo,
        CategoryInfo,
        DateInfo,
        OriginalInfo,
        PageViewInfo: SUPPORT_PAGEVIEW ? PageViewInfo : () => null,
        ReadingTimeInfo: ENABLE_READING_TIME ? ReadingTimeInfo : () => null,
        TagInfo,
        WordInfo,
    },
    props: {
        /**
         * Article information to display
         *
         * 待展示的文章信息
         */
        items: {
            type: [Array, Boolean],
            default: () => [
                "Author",
                "Original",
                "Date",
                "Category",
                "Tag",
                "ReadingTime",
            ],
        },
        /**
         * Article information
         *
         * 文章信息配置
         */
        info: {
            type: Object,
            required: true,
        },
    },
    setup(props) {
        const pure = usePure();
        return () => props.items
            ? h("div", { class: "page-info" }, props.items.map((item) => h(resolveComponent(`${item}Info`), {
                ...props.info,
                pure: pure.value,
            })))
            : null;
    },
});
